libs/async/include/hpx/async/async_callback.hpp

```diff
@@ -0,0 +1,276 @@
+            typename Derived, typename Callback, typename... Ts>
+        HPX_FORCEINLINE static lcos::future<
+            typename traits::promise_local_result<typename traits::
+                    extract_action<Derived>::remote_result_type>::type>
+        call(Policy_&& launch_policy,
+            hpx::actions::basic_action<Component, Signature, Derived> const&,
+            naming::id_type const& id, Callback&& cb, Ts&&... ts)
+        {
+            return async_cb<Derived>(std::forward<Policy_>(launch_policy), id,
+                std::forward<Callback>(cb), std::forward<Ts>(ts)...);
+        }
+
+        template <typename Policy_, typename Component, typename Signature,
+            typename Derived, typename Client, typename Stub, typename Callback,
+            typename... Ts>
+        HPX_FORCEINLINE static lcos::future<
+            typename traits::promise_local_result<typename traits::
+                    extract_action<Derived>::remote_result_type>::type>
+        call(Policy_&& launch_policy,
+            hpx::actions::basic_action<Component, Signature, Derived> const&,
+            components::client_base<Client, Stub> const& c, Callback&& cb,
+            Ts&&... ts)
+        {
+            typedef typename components::client_base<Client,
+                Stub>::server_component_type component_type;
+
+            typedef traits::is_valid_action<Derived, component_type> is_valid;
+            static_assert(is_valid::value,
+                "The action to invoke is not supported by the target");
+
+            return async_cb<Derived>(std::forward<Policy_>(launch_policy),
+                c.get_id(), std::forward<Callback>(cb),
+                std::forward<Ts>(ts)...);
+        }
+
+        template <typename Policy_, typename Component, typename Signature,
+            typename Derived, typename DistPolicy, typename Callback,
+            typename... Ts>
+        HPX_FORCEINLINE static lcos::future<
+            typename traits::promise_local_result<typename traits::
+                    extract_action<Derived>::remote_result_type>::type>
+        call(Policy_&& launch_policy,
+            hpx::actions::basic_action<Component, Signature, Derived> const&,
+            DistPolicy const& policy, Callback&& cb, Ts&&... ts)
+        {
+            return async_cb<Derived>(std::forward<Policy_>(launch_policy),
+                policy, std::forward<Callback>(cb), std::forward<Ts>(ts)...);
+        }
+    };
+}}    // namespace hpx::detail
+
+namespace hpx {
+    template <typename F, typename... Ts>
+    HPX_FORCEINLINE auto async_cb(F&& f, Ts&&... ts) -> decltype(
+        detail::async_cb_dispatch<typename util::decay<F>::type>::call(
+            std::forward<F>(f), std::forward<Ts>(ts)...))
+    {
+        return detail::async_cb_dispatch<typename util::decay<F>::type>::call(
+            std::forward<F>(f), std::forward<Ts>(ts)...);
+    }
+}    // namespace hpx
```

Comment:
Does this API function belong into the `local_async` module?

Reply:
I'm not sure. `async_cb_dispatch` only has `call` overloads for actions (AFAICT). Is it meant to have local dispatch points as well? We can of course still move `async_cb` itself into the `local_async` module in case `async_cb` for local tasks ever gets implemented.

Reply:
Fair points. I didn't look at the dispatch points, so I think we're fine with what you propose.